sdk/keyvault/keyvault-keys/src/localCryptography/algorithms.browser.ts

```diff
@@ -0,0 +1,174 @@
+
+import { isNode } from "@azure/core-http";
+import { JsonWebKey, KeyOperation } from "../keysModels";
+import { LocalCryptographyUnsupportedError } from "./models";
+
+/**
+ * This file contains the implementation of local supported algorithms, which
+ * consist of wrappers around methods already available as part of NodeJS.
+ *
+ * The goal of this file is to abstract away how we implement these algorithms,
+ * since we might be using environment-specific code or dependencies.
+ *
+ * Some algorithms can't be implemented neither in the Browser or in Node
+ * without third party dependencies or considerable development.
+ *
+ * We currently don't support any operation that requires the private data of a
+ * key. Once we implement support for the private part of JWK keys and PEM keys,
+ * we will be able to increase the support of our existing algorithms.
+ */
+
+/**
+ * @internal
+ * @ignore
+ * Abstract representation of a assertion.
+ * Assertions verify that the requirements to execute a local cryptography operation are met.
+ * @param key The JSON Web Key that will be used during the local operation.
+ * @param operationName The name of the operation, as in "encrypt", "decrypt", "sign", etc.
+ */
+export type LocalAssertion = (
+  key?: JsonWebKey,
+  operationName?: LocalCryptographyOperationName
+) => void;
+
+/**
+ * @internal
+ * @ignore
+ * The list of known assertions so far.
+ * Assertions verify that the requirements to execute a local cryptography operation are met.
+ */
+export const assertions: Record<"keyOps" | "rsa" | "nodeOnly", LocalAssertion> = {
+  /**
+   * Validates that the target local cryptography operation is allowed by the key's "keyOps" property.
+   */
+  keyOps(key?: JsonWebKey, operationName?: LocalCryptographyOperationName): void {
+    if (key && key.keyOps && !key.keyOps.includes(operationName as KeyOperation)) {
+      throw new Error(`Key does not support the ${operationName} operation`);
+    }
+  },
+  /**
+   * Validates that the key type is "RSA". For operations on RSA keys.
+   */
+  rsa(key?: JsonWebKey): void {
+    if (key && key.kty! !== "RSA" && key.kty! !== "RSA-HSM") {
+      throw new Error("Key type does not match the algorithm RSA");
+    }
+  },
+  /**
+   * Passes only when the operation is being executed in NodeJS.
+   * For operations that can only run in NodeJS.
+   */
+  nodeOnly(): void {
+    if (!isNode) {
+      throw new LocalCryptographyUnsupportedError("This operation is only available in NodeJS");
+    }
+  }
+};
+
+/**
+ * TypeScript fancy for making plain objects require at least one key-value pair of another set of key-values.
+ */
+export type RequireAtLeastOne<T> = {
+  [K in keyof T]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<keyof T, K>>>;
+}[keyof T];
+
+/**
+ * Union type representing the names of the supported local cryptography operations.
+ */
+export type LocalCryptographyOperationName = "encrypt" | "wrapKey" | "createHash" | "verify";
```

Comment:
The idea behind having a separate file for browser counterpart to code is to house the difference in implementations between browser and node. 

There are quite a few types in this file which are same as in algorithms.ts file. `LocalCryptographyOperationName` is one example.

These should be moved to a common file and re-used from both algorithms.ts and algorithms.browser.ts file instead having the same type defined twice.

Or alternatively, we can focus on what is different between node and browser, and create a new file to keep the differences. From where I can see, the main point of difference is the contents of `localSupportedAlgorithms`. Can we consider the below?
- Delete the algorithms.browser.ts file and create new files `localAlgorithms.ts` and `localAlgorithms.browser.ts` instead
- Move the `localSupportedAlgorithms` constant and the code required to fill it up from `algorithms.ts` to `localAlgorithms.ts` file.
- The `localAlgorithms.browser.ts` will also export the constant `localSupportedAlgorithms`, but it will be empty because we don't have any local supported algorithms in the browser. This file need not have anything else.

Reply:
Can do!

Reply:
So, here's what I did:

- I didn't decide to name these files `localAlgorithms` because it's already in the folder `localCryptography`.
- I moved all of the common types between `algorithms.ts` and `algorithms.browser.ts` to `models.ts`. This file already existed and used to contain only one class, `LocalCryptographyUnsupportedError`.
- After moving the common types, `algorithms.browser.ts` is empty, as you suggested. More properly, it only has a function, `isLocallySupported`, that always returns false, and an empty list of `localSupportedAlgorithms`, for parity with the other file.
- Ensured everything worked (linting, tests, etc).

Once we decide to invest time adding support for browser algorithms, we'll be able to re-use these changes and do further improvements. Let me know if this moves this PR closer to a good state 🌞